### ISSUE TITLE
Problem with sayall - TTS not playing

### ISSUE THREAD

**Issue by dolphinkite** [jishi/node-sonos-http-api#208]
Hi,

Still trying to get my doorbell feature to work. After re-reading the readme, I now understand that I do NOT need to do "pauseall" nor "resumeall" when someone rings the doorbell. The action "sayall" should already do that - pause, group, play, ungroup, unpause; correct?

It is not working properly for me. Here is my scenario:
Player 1 + Player 2 already grouped - playing slacker radio
Player 3: Not playing anything
Player 4 : Playing local playlist
Player 5 : Playing different playlist

I tried to trigger the sayall function with a TTS phrase, but here are my observations:
1- All player are indeed paused and grouped and the downloaded mp3 is queued
2- For some reason, the downloaded and queued mp3 doesn't play
3- If I manually "play" it, then once it is played, the zones will be ungrouped and each player will be started if it was before.

In other words, it sorta works, except the "play" part.

What can I do to help debug this?

PS: The latency on the operation is pretty high.. not sure it will work well given how much time it takes. But still a fun feature.

**Comment by jishi** [comment]
Hi. Yeah, the sayall will as you mentioned, pause and group all players. I just didn't think that was what you wanted :)

It has some quirks, it will not restore radios correctly, for instance, and it doesn't always manage with the timing unfortunately. And as you've noticed, the Sonos system isn't always super quick with it's operation, especially not if you have high wifi congestion. 

I spent many, many nights to get it working, and even then, it wasn't very reliable. It has always been flagged as an experimental feature, and I have had other priorities. I'm hoping to revisit it after my promisification of the API, which would make it easier to build it the way I want, and it will also require some retry logic to be resilient enough to temporary glitches.

**Comment by jishi** [comment]
Could you please try out the beta branch? The sayall command there is vastly improved (although still issues with radio stations).

**Comment by dolphinkite** [comment]
Hi,

Despite my effort, I can't get the beta branch to work. As I said earlier, I am on Docker. When I try to create a container I get the following errors in the logs. I understand from the first line that the settings.json is missing, but I must first create the container, and THEN I copy over the .json file. but it seems that right from the start, it is crashing before I can copy anything over.

Also to note, when I did the first build of the docker image, I got the following error message: `npm WARN cls-bluebird@1.0.1 requires a peer of continuation-local-storage@~3 but none was installed.`

See log file attached
[docker log.txt](https://github.com/jishi/node-sonos-http-api/files/287275/docker.log.txt)

**Comment by dolphinkite** [comment]
I managed to get rid of the first errors (missing settings.json file) by using the "docker create" command and then copying the settings.json file before starting the container (create, as opposed to run, will create the container but not start it). The new log is identical to the previous one, with the same problem: `/opt/app/node_modules/sonos-discovery/lib/models/Player.js:89
      track.duration = parseTime(item.res.$attrs.duration);
                                                ^

TypeError: Cannot read property 'duration' of undefined
    at EventEmitter.<anonymous> (/opt/app/node_modules/sonos-discovery/lib/models/Player.js:89:49)
    at emitOne (events.js:90:13)
    at EventEmitter.emit (events.js:182:7)
    at SAXStream.onCloseTag (/opt/app/node_modules/xml-flow/lib/xml-flow.js:131:15)
    at emitOne (events.js:90:13)
    at SAXStream.emit (events.js:182:7)
    at Object.me._parser.(anonymous function) [as onclosetag] (/opt/app/node_modules/sax/lib/sax.js:258:17)
    at emit (/opt/app/node_modules/sax/lib/sax.js:640:35)
    at emitNode (/opt/app/node_modules/sax/lib/sax.js:645:5)
    at closeTag (/opt/app/node_modules/sax/lib/sax.js:905:7)
    at Object.write (/opt/app/node_modules/sax/lib/sax.js:1449:13)
    at SAXStream.write (/opt/app/node_modules/sax/lib/sax.js:239:18)
    at Readable.ondata (_stream_readable.js:529:20)
    at emitOne (events.js:90:13)
    at Readable.emit (events.js:182:7)
    at readableAddChunk (_stream_readable.js:147:16)

`

**Comment by jishi** [comment]
What is the state of the players (what source etc)  when this happens?

**Comment by dolphinkite** [comment]
Ok, so here is the scenario (from what I remember):
At first, I tried the "ultimate" scenario. 5 zones playing 5 different sources: slacker, local playlist, tune-in, Google Music being chromecasted from my phone and something else. It started, I was able to obtain the zones, then I tried the "Sayall" command and the mp3 stating that I was missing the key played. Then I tried to add the key file but nothing was working (I may, in the mean time, change what was playing...). I tried to stop and remove the container and create it again but it didn't work. I tried to stop all sources, but was still getting the same error message.

This morning, I tried to change all sources to only local library and it works fine, whether they are playing or not.

**Comment by jishi** [comment]
Hm, okay. Some source seems to render XML metadata that isn't parsed correctly. I added a fix for it in beta.13, update to the latest commit on the beta branch, and update the dependencies with `npm install --production`

This version also includes a preliminary fix for restoring radio stations during say/sayall commands. Unfortunately it can't restore to the same song it was playing if it is a spotify radiostation (and I guess pandora and similar services would suffer the same issue), but Tunein is fine. Please try it out and let me know your experience.

**Comment by dolphinkite** [comment]
Preliminary testing (1 zone with tune-in) shows success. I'll do more tests tomorrow when I do not risk waking the whole household.

Thanks

**Comment by jishi** [comment]
Closing for now, reopen if problem arises again

**Comment by dolphinkite** [comment]
Ok, the simple scenario works, but more complex setup don't.

Scenario 1:
Zone 1: Tune-in
Zone 2: Skacker
Zone 3: Local library
Zone 4: Local Library
Zone 5: Local Library
==> All zones get grouped, the TTS is played (I can't really verify that the TTS plays on all zone though), all zones get ungrouped, Z1, Z3 and Z4 resume playing (tune-in, local library) but Z2 doesn't resume playing (but Slacker is selected as the source) and Z5 also does not resume playing (but the TTS mp3 is still in the queue).

Scenario 2:
Zone 1: Tune-In
Zone 2: Slacker
Zone 3: Local library
Zone 4: Google Music being Chromecasted from my phone
Zone 5: Local library

==> All zones get grouped, the TTS is played (I can't really verify that the TTS plays on all zone though), Z4 is ungrouped with an empty queue, zones 1, 2, 3, 5 are still grouped, nothing resumes playing.

Let me know if I can share a log file with you.

**Comment by jishi** [comment]
Something probably broke when restoring Z4, hence Z5 wasn't restored correctly.

Could you start the server with trace logging activated? on linux/OS X: 

`NODE_LOG_LEVEL=trace npm start > trace.log`

It is enough to just run slacker on one zone. After it breaks, stop the server and email the log file to me jayjay@grabbarna.se

**Comment by dolphinkite** [comment]
Are we testing scenario 1 or 2? The difference being the addition of a zone being Chromecasted.

I tried to simply have 2 zones playing slacker and the other zones not playing anything.. and everything was restored correctly. 

Because I'm in docker, I'm not sure how to start the container with the params above, but I'll email you my logs (with the docker logs command) and we'll see if you have when you need.

**Comment by jishi** [comment]
Ah sorry, read too quick. Preferably both, I guess both Google Music and Slacker has some issues (since Z2 was still grouped in scenario 2). But we might want to fix one of the services at a time. To simplify the test, just make sure one zone is using the offending music services, not necessarily needed to run all 5 zones.
